### PR TITLE
Image Embedder C Tasks API 

### PR DESCRIPTION
**Pull request by kinaryml**
- Add the files necessary to implement the Image Embedder C API
- Add a utility to compute similarity between two embeddings (for text and vision tasks)

### CODE REVIEW

File: mediapipe/tasks/c/components/containers/embedding_result_converter.cc [hunk 1]
```diff
@@ -66,6 +66,29 @@ void CppConvertToEmbeddingResult(
...
+  if (in.quantized_embedding != nullptr) {
+    out->quantized_embedding.assign(in.quantized_embedding,
+                                    in.quantized_embedding + in.values_count);
+  }
+
+  out->head_index = in.head_index;
+
+  // Copy head_name if it is present.
+  if (in.head_name) {
+    out->head_name = std::make_optional(std::string(in.head_name));
```
Comment (schmidt-sebastian) [inline mediapipe/tasks/c/components/containers/embedding_result_converter.cc:88]: Make_optional shouldn't be needed as this should be converted during assignment.

File: mediapipe/tasks/c/components/containers/embedding_result_converter.h [hunk 2]
```diff
@@ -29,6 +29,10 @@ void CppConvertToEmbeddingResult(
     const mediapipe::tasks::components::containers::EmbeddingResult& in,
     EmbeddingResult* out);
 
+void ConvertToCppEmbedding(
```
Comment (schmidt-sebastian) [inline mediapipe/tasks/c/components/containers/embedding_result_converter.h:32]: This is a C++ function so it should start with Cpp, even though `CppConvertToCppEmbedding` is not going to win us any awards :)

File: mediapipe/tasks/c/text/text_embedder/text_embedder.h [hunk 3]
```diff
@@ -66,6 +66,15 @@ MP_EXPORT void text_embedder_close_result(TextEmbedderResult* result);
 // allocated for the error message.
 MP_EXPORT int text_embedder_close(void* embedder, char** error_msg);
 
+// Utility function to compute cosine similarity [1] between two embeddings.
+// May return an InvalidArgumentError if e.g. the embeddings are of different
+// types (quantized vs. float), have different sizes, or have a an L2-norm of
+// 0.
+//
+// [1]: https://en.wikipedia.org/wiki/Cosine_similarity
+MP_EXPORT int cosine_similarity(const Embedding& u, const Embedding& v,
```
Comment (schmidt-sebastian) [inline mediapipe/tasks/c/text/text_embedder/text_embedder.h:75]: Think we should prefix this with "text_embedder_" so we don't get duplicate symbol errors.

File: mediapipe/tasks/c/vision/image_embedder/image_embedder.h [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+#define MP_EXPORT __attribute__((visibility("default")))
+#endif  // MP_EXPORT
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+typedef EmbeddingResult ImageEmbedderResult;
+
+// Supported image formats.
```
Comment (schmidt-sebastian) [inline mediapipe/tasks/c/vision/image_embedder/image_embedder.h:35]: This needs to be moved to a common file for all image tasks.

File: mediapipe/tasks/c/vision/image_embedder/image_embedder.h [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+
+// Performs embedding extraction on the input `image`. Returns `0` on success.
+// If an error occurs, returns an error code and sets the error parameter to an
+// an error message (if `error_msg` is not `nullptr`). You must free the memory
+// allocated for the error message.
+MP_EXPORT int image_embedder_embed_image(void* embedder, const MpImage* image,
+                                         ImageEmbedderResult* result,
+                                         char** error_msg);
+
+MP_EXPORT int image_embedder_embed_for_video(void* embedder,
```
Comment (schmidt-sebastian) [inline mediapipe/tasks/c/vision/image_embedder/image_embedder.h:124]: We should add comments here (and in the image classifier as well).

File: mediapipe/tasks/c/vision/image_embedder/image_embedder.h [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+// allocated for the error message.
+MP_EXPORT int image_embedder_close(void* embedder, char** error_msg);
+
+// Utility function to compute cosine similarity [1] between two embeddings.
+// May return an InvalidArgumentError if e.g. the embeddings are of different
+// types (quantized vs. float), have different sizes, or have a an L2-norm of
+// 0.
+//
+// [1]: https://en.wikipedia.org/wiki/Cosine_similarity
+MP_EXPORT int cosine_similarity(const Embedding& u, const Embedding& v,
```
Comment (schmidt-sebastian) [inline mediapipe/tasks/c/vision/image_embedder/image_embedder.h:150]: Prefix with image_embedder to avoid duplicate symbol names.

**Review by schmidt-sebastian — CHANGES_REQUESTED** [review]

**Comment by kinaryml** [comment]
@schmidt-sebastian I've resolved those comments. PTAL.

**Review by schmidt-sebastian — APPROVED** [review]

**Pull request merged**